modules/core/src/test/java/org/apache/ignite/cache/ReleaseSegmentOnHistoricalRebalance.java

```diff
@@ -0,0 +1,408 @@
+        checkHistoricalRebalance(n -> {
+            IgniteInternalCache cachex = n.cachex(DEFAULT_CACHE_NAME);
+
+            GridCacheOffheapManager spy = spy(offheapMgr(cachex));
+
+            doAnswer(m -> {
+                CheckpointHistory cpHist = dbMgr(n).checkpointHistory();
+
+                for (Long cp : cpHist.checkpoints())
+                    release(n, entry(cpHist, cp).checkpointMark());
+
+                return m.callRealMethod();
+            }).when(spy).rebalanceIterator(any(), any());
+
+            offheapMgr(cachex, spy);
+        });
+    }
+
+    /**
+     * Checks that if release the segment during {@link IgniteCacheOffheapManagerImpl#rebalanceIterator},
+     * there will be no errors and the rebalance will be completed.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testReleaseDuringRebalanceIterator() throws Exception {
+        checkHistoricalRebalance(n -> {
+            IgniteInternalCache cachex = n.cachex(DEFAULT_CACHE_NAME);
+
+            GridCacheOffheapManager spy = spy(offheapMgr(cachex));
+
+            doAnswer(m -> {
+                WALPointer reserved = dbMgr(n).latestWalPointerReservedForPreloading();
+
+                assertNotNull(reserved);
+
+                Object o = m.callRealMethod();
+
+                release(n, reserved);
+
+                assertTrue(segmentAware(n).minReserveIndex(Long.MAX_VALUE));
+
+                return o;
+            }).when(spy).rebalanceIterator(any(), any());
+
+            offheapMgr(cachex, spy);
+        });
+    }
+
+    /**
+     * Checks that if the reservation is released immediately,
+     * there will be no errors and the rebalance will be completed.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testImmediateReleaseSegment() throws Exception {
+        checkHistoricalRebalance(n -> {
+            SegmentAware spy = spy(segmentAware(n));
+
+            doAnswer(m -> {
+                Object o = m.callRealMethod();
+
+                spy.release(m.getArgument(0));
+
+                return o;
+            }).when(spy).reserve(anyLong());
+
+            segmentAware(n, spy);
+        });
+    }
+
+    /**
+     * Checks that that if there is no reservation,
+     * there will be no errors and the rebalance will be completed.
+     *
+     * @throws Exception If failed.
+     */
+    @Test
+    public void testNoReserveSegment() throws Exception {
+        checkHistoricalRebalance(n -> {
+            SegmentAware spy = spy(segmentAware(n));
+
+            when(spy.reserve(anyLong())).thenAnswer(m -> false);
+
+            segmentAware(n, spy);
+        });
+    }
+
+    /**
+     * Populate the cache.
+     *
+     * @param cache Cache.
+     * @param cnt Entry count.
+     * @param o Key offset.
+     * @throws Exception If failed.
+     */
+    private void populate(IgniteCache<? super Object, ? super Object> cache, int cnt, int o) throws Exception {
```

Comment:
well, the key offset for the type `? super Object` looks odd to me :) Perhaps, it can be changed to `IgniteCache<Integer, ? super Object> cache`. WDYT?